presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java

```diff
@@ -214,9 +220,14 @@ public MongoClientConfig setSocketKeepAlive(boolean socketKeepAlive)
         return this;
     }
 
-    public ReadPreferenceType getReadPreference()
+    public ReadPreference getReadPreference()
     {
-        return readPreference;
+        if (readPreferenceTagSets.isEmpty()) {
+            return readPreference.getReadPreference();
```

Comment:
maybe rename the ReadPreferenceType field from readPreference to defaultReadPreference since that's not necessarily the ReadPreference that's going to be used.

Reply:
@rschlussel , Thank you for reviewing the codes. The readPreference is other configure set by mongodb.read-preference that have default value-primary. If tag sets are empty, the readPreference is used directly. If not, TaggableReadPreference with tag sets and readPreference. 

In addition, I do want to add test code for this. However, I think that it's impossible to test read preference in fake mongodb server of test code. In fact, I was forced to test this through real mongodb replica set. Could I have any advice for this?

Reply:
To clarify I mean that instead of calling `readPreference.getReadPreference()` you would call something like`readPreference.getDefault()`.  The goal is to indicate that the specific readPreference object isn't necessarily the one that's going to be used because if you've specified tags you'll create a new TaggableReadPreference.

Have you looked at the MongoQueryRunner? Can you explain why testing with testing with the Mongo query runner doesn't work? 

@ebyhr do you have any suggestions?

Reply:
I suppose that the method should return readPreference value set by the configure-"mongodb.read-preference" not default although there arent's tagsets for read-preference.

The properties of "read-preference" and "read-preference-tags" are different configures, but MongoClientOptions.Builder takes a parameter as com.mongodb.ReadPreference class that includes two configures. So, I modified the method to return ReadPreference with tagsets if tagsets exist. This makes be confusing :(

How about below codes?

MongoClientConfig.java
```
    ...
    //rollback to previous codes
    public ReadPreferenceType getReadPreference()
    {
        return readPreference;
    }

    @Config("mongodb.read-preference")
    public MongoClientConfig setReadPreference(ReadPreferenceType readPreference)
    {
        this.readPreference = readPreference;
        return this;
    }

    ...

    public List<TagSet> getReadPreferenceTags()
    {
        return readPreferenceTagSets;
    }

    @Config("mongodb.read-preference-tags")
    public MongoClientConfig setReadPreferenceTags(String readPreferenceTags)
    {
        this.readPreferenceTagSets = buildTagSets(TAGSET_SPLITTER.split(readPreferenceTags));
        return this;
    }
    ...
```

MongoClientModule.java
```
....
    @Singleton
    @Provides
    public static MongoSession createMongoSession(TypeManager typeManager, MongoClientConfig config)
    {
        requireNonNull(config, "config is null");

        MongoClientOptions.Builder options = MongoClientOptions.builder();

        options.connectionsPerHost(config.getConnectionsPerHost())
                .connectTimeout(config.getConnectionTimeout())
                .socketTimeout(config.getSocketTimeout())
                .socketKeepAlive(config.getSocketKeepAlive())
                .sslEnabled(config.getSslEnabled())
                .maxWaitTime(config.getMaxWaitTime())
                .minConnectionsPerHost(config.getMinConnectionsPerHost())
                .readPreference(getReadPreference(config))
                .writeConcern(config.getWriteConcern().getWriteConcern());

        if (config.getRequiredReplicaSetName() != null) {
            options.requiredReplicaSetName(config.getRequiredReplicaSetName());
        }

        MongoClient client = new MongoClient(config.getSeeds(), config.getCredentials(), options.build());

        return new MongoSession(
                typeManager,
                client,
                config);
    }

    private ReadPreference getReadPreference(MongoClientConfig config)
    {
        if (config.getReadPreferenceTags().isEmpty()) {
            return config.getReadPreadPreference.getReadPreference();
        }
        else {
            return ReadPreference.valueOf(config.getReadPreadPreference.name(), config.getReadPreferenceTags());
        }
    }

 ...
```

Reply:
Also, can you check what the exception looks like to the user if you try to specify tags and the ReadPreference is set to Primary?  You may need to catch/wrap that exception as a PrestoException so that the user can get a friendlier error message.